cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,331 @@
+
+        QSize itemSize = item->sizeHint(); // The suggested size for the item.
+        const int itemWidth = itemSize.width() + horizontalSpacing();
+
+        // Check if the item fits in the current row's remaining width.
+        if (currentXPosition + itemWidth > availableWidth) {
+            // If not, layout the current row and start a new row.
+            layoutSingleRow(rowItems, originX, currentYPosition);
+            rowItems.clear();                                  // Clear the temporary storage for the new row.
+            currentXPosition = originX;                        // Reset x-position to the start of the new row.
+            currentYPosition += rowHeight + verticalSpacing(); // Move y-position down for the new row.
+            rowHeight = 0;                                     // Reset row height for the new row.
+        }
+
+        // Add the item to the current row.
+        rowItems.append(item);
+        rowHeight = qMax(rowHeight, itemSize.height());             // Update the row height to the tallest item.
+        currentXPosition += itemSize.width() + horizontalSpacing(); // Move x-position for the next item.
+    }
+
+    // Layout the final row if there are remaining items.
+    layoutSingleRow(rowItems, originX, currentYPosition);
+
+    // Return the total y-coordinate at the bottom of the last row.
+    return currentYPosition + rowHeight;
+}
+
+/**
+ * @brief Helper function for arranging a single row of items within specified bounds.
+ * @param rowItems Items to be arranged in the row.
+ * @param x The x-coordinate for starting the row.
+ * @param y The y-coordinate for starting the row.
+ */
+void FlowLayout::layoutSingleRow(const QVector<QLayoutItem *> &rowItems, int x, const int y)
+{
+    // Iterate through each item in the row and position it.
+    for (QLayoutItem *item : rowItems) {
+        if (item == nullptr || item->isEmpty()) {
+            continue;
+        }
+        QSize itemMaxSize = item->widget()->maximumSize(); // Get the item's maximum allowable size.
+        // Constrain the item's width and height to its size hint or maximum size.
+        int itemWidth = qMin(item->sizeHint().width(), itemMaxSize.width());
+        int itemHeight = qMin(item->sizeHint().height(), itemMaxSize.height());
+        // Set the item's geometry based on the calculated size and position.
+        item->setGeometry(QRect(QPoint(x, y), QSize(itemWidth, itemHeight)));
+        // Move the x-position for the next item, including horizontal spacing.
+        x += itemWidth + horizontalSpacing();
+    }
+}
+
+/**
+ * @brief Returns the preferred size for this layout.
+ * @return The maximum of all item size hints as a QSize.
+ */
+QSize FlowLayout::sizeHint() const
+{
+    QSize size;
+    for (const QLayoutItem *item : items) {
+        if (!(item == nullptr || item->isEmpty())) {
+            size = size.expandedTo(item->sizeHint());
+        }
+    }
+    return size.isValid() ? size : QSize(0, 0);
```

Comment:
I think you can return {0,0} for the default case

Reply:
It wants an expression.